File: leetcode/medium/142.Linked List Cycle II/js/linkedListCycleII.js
"use strict";

// Linked List Cycle

/**
 * 
 * @param {ListNode} head 
 */
// Approach 1: HashTable
var linkedListCycle = function(head) {
    let mySet = new Set();
    while (head !== null) {
        if (mySet.has(head)) {
            return true;
        };
        mySet.add(head);
        head = head.next;
    };
    return false;
};

// Complexity Analysis:
// Time Complexity: O(N)
// Space Complexity: O(N)

// Approach 2: Floyd's Cycle Finding Algorithm

/**
 * 
 * @param {ListNode} head 
 */
var hasCycle = function(head) {
    let tortoise = head;
    let hare = head;
    while (hare !== null && hare.next !== null) {
        tortoise = tortoise.next;
        hare = hare.next.next;
        if (tortoise === hare) {
            hare = head;
            while (hare !== tortoise) {
                hare = hare.next;
                tortoise = tortoise.next;
            };
            return tortoise;
        };
    };
    return null;
};
// Complexity Analysis:
// Time Complexity: O(N)
// Space Complexity: O(1)

File: leetcode/medium/142.Linked List Cycle II/python/linkedListCycleII.py
"""Linked List Cycle II"""

# Approach 1. Hash Table
def detect_cycle(head):
    my_set = set()
    while head is not None:
        if head in my_set:
            return head
        my_set.add(head)
        head = head.next
    return None

'''Complexity Analysis:
Time Complexity: O(N)
Space Complexity: O(N)'''

# Approach 2: Floyd's Tortoise and Hare Algorithm
def detect_cycle(head):
    tortoise = head
    hare = head
    while hare and hare.next is not None:
        tortoise = tortoise.next
        hare = hare.next.next
        if hare == tortoise:
            hare = head
            while hare != tortoise:
                hare = hare.next
                tortoise = tortoise.next
            return tortoise
    return None
    

'''Complexity Analysis:
Time Complexity: O(N)
Space Complexity: O(1)'''
